lib/Parse/ParseDecl.cpp

```diff
@@ -3236,6 +3309,25 @@ bool Parser::parseDeclList(SourceLoc LBLoc, SourceLoc &RBLoc,
   return !RBLoc.isValid();
 }
 
+bool Parser::canDelayBodyParsing() {
+  if (SF.Kind == SourceFileKind::REPL)
+    return false;
+  if (SF.shouldCollectToken())
+    return false;
+  if (SF.shouldBuildSyntaxTree())
```

Comment:
Do we have to disable this optimization when building the syntax tree? It seems like building the syntax tree would effectively parse everything in the source file for which we're building the syntax tree, but otherwise is fine?

Reply:
The underlying concept of lazy parsing should also apply for SyntaxTree. The reason we disable it here is that we always visit the whole syntax tree and sent it over to the editor side. Thus, in practice, syntax tree parsing cannot benefit from the laziness as much as AST. I'll add some comments here to explain.